src/selector_gui.cpp

```diff
@@ -0,0 +1,383 @@
+ * Event handler
+ *
+ * All of these parameters should be passed straight from the parent's window's OnResize
+ * Event handler, without any changes, or conditional checks
+ *
+ * @see Window::OnResize()
+ */
+void SelectorWidget::OnResize()
+{
+	this->vscroll->SetCapacityFromWidget(parent_window, WID_SELECTOR_MATRIX);
+}
+
+/**
+ * Selector widget's OnEditboxChanged event handler
+ * This function is meant to be called from the parent's window's OnEditboxChanged
+ * Event handler
+ *
+ * @param wid The widget id of the editbox.
+ *
+ * All of these parameters should be passed straight from the parent's window's OnEditboxChanged
+ * Event handler, without any changes, or conditional checks
+ *
+ * @see Window::OnEditboxChanged()
+ */
+void SelectorWidget::OnEditboxChanged(WidgetID wid)
+{
+	if (wid != WID_SELECTOR_EDITBOX) return;
+
+	this->string_filter.SetFilterTerm(this->editbox.text.buf);
+	this->parent_window->InvalidateData(0);
+	this->vscroll->SetCount(this->filtered_list.size());
+}
+
+/**
+ * Selector widget's DrawWidget method
+ * This function is meant to be called from the parent's window's DrawWidget
+ * method
+ *
+ * @param widget  Widget number.
+ * @param[In,out] size Size of the widget.
+ * @param padding Recommended amount of space between the widget content and the widget edge.
+ * @param[In,out] fill Fill step of the widget.
+ * @param[In,out] resize Resize step of the widget.
+ *
+ * All of these parameters should be passed straight from the parent's window's DrawWidget
+ * Event handler, without any changes, or conditional checks
+ *
+ * This function calls a profile defined function this->profile.DrawSection(),
+ * that actually draws the fields
+ *
+ * @see Window::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+void SelectorWidget::DrawWidget(const Rect &r, WidgetID widget)
+{
+	if (widget != WID_SELECTOR_MATRIX) return;
+
+	Rect line = r.WithHeight(row_height);
+
+	auto [first, last] = this->vscroll->GetVisibleRangeIterators(filtered_list);
+
+	for (auto it = first; it != last; ++it) {
+		const Rect ir = line.Shrink(WidgetDimensions::scaled.framerect);
+
+		if (this->shown[*it]) {
+			DrawFrameRect(line, COLOUR_BROWN, FR_LOWERED);
+		}
+		this->DrawSection(*it, ir.Shrink(WidgetDimensions::scaled.matrix));
+
+		line = line.Translate(0, row_height);
+	}
+}
+
+/**
+ * A function that updates and rebuilds the list of selectable items
+ * This function calls a profile defined function this->profile.RebuildList(),
+ * that actually rebuilds the list of items.
+ *
+ * It is also manages the shown items and the filtered_list lists
+ *
+ * @see SelectorWidget::Profile
+ */
+void SelectorWidget::RebuildList()
+{
+	this->list.clear();
+	this->filtered_list.clear();
+	this->PopulateList();
+	uint32_t max = *std::max_element(std::begin(this->list), std::end(this->list));
+
+	for (size_t i = this->shown.size(); i <= max ; i++) {
+		this->shown.push_back(true);
+	}
```

Comment:
Maybe include `this->shown.reserve(max);` first to avoid the vector being reallocated.

Reply:
I will leave this for @PeterN to resolve.